substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -0,0 +1,1527 @@
+								_ => {
+									log::error!(target: LOG_TARGET, "[{:?}] txpool: update_view: something went wrong: {error:?}", tx_hash);
+									Err((
+										Error::UnknownTransaction(UnknownTransaction::CannotLookup),
+										tx_hash,
+										tx.tx(),
+									))
+								},
+							}
+						},
+						|watcher| Ok(watcher),
+					);
+
+					if let Ok(watcher) = result {
+						log::trace!(target: LOG_TARGET, "[{:?}] adding watcher {:?}", tx_hash, view.at.hash);
+						self.view_store
+							.listener
+							.add_view_watcher_for_tx(
+								tx_hash,
+								view.at.hash,
+								watcher.into_stream().boxed(),
+							);
+						Ok(())
+					} else {
+						result.map(|_| ())
+					}
+				}
+			})
+			.collect::<Vec<_>>();
+
+		let results = future::join_all(results).await;
+		let submitted_count = submitted_count.load(Ordering::Relaxed);
+
+		log::info!(target: LOG_TARGET, "update_view_pool: at {:?} watched {}/{}", view.at.hash, submitted_count, self.mempool_len().1);
+
+		all_submitted_count += submitted_count;
+		let _ = all_submitted_count
+			.try_into()
+			.map(|v| self.metrics.report(|metrics| metrics.submitted_from_mempool_txs.inc_by(v)));
+
+		// if there are no views yet, and a single newly created view is reporting error, just send
+		// out the invalid event, and remove transaction.
+		if self.view_store.is_empty() {
+			for result in results {
+				match result {
+					Err((Error::InvalidTransaction(_), tx_hash, tx)) => {
+						self.view_store.listener.invalidate_transactions(vec![tx_hash]);
+						self.mempool.remove_watched(&tx);
+					},
+					_ => {},
+				}
+			}
+		}
+	}
+
+	async fn update_view_with_fork(
+		&self,
+		view: &View<ChainApi>,
+		tree_route: &TreeRoute<Block>,
+		hash_and_number: HashAndNumber<Block>,
+	) {
+		log::debug!(target: LOG_TARGET, "update_view_with_fork tree_route: {:?} {tree_route:?}", view.at);
+		let api = self.api.clone();
+
+		// We keep track of everything we prune so that later we won't add
+		// transactions with those hashes from the retracted blocks.
+		let mut pruned_log = HashSet::<ExtrinsicHash<ChainApi>>::new();
+
+		future::join_all(
+			tree_route
+				.enacted()
+				.iter()
+				// .chain(std::iter::once(&hash_and_number))
+				.map(|h| crate::prune_known_txs_for_block(h, &*api, &view.pool)),
+		)
+		.await
+		.into_iter()
+		.for_each(|enacted_log| {
+			pruned_log.extend(enacted_log);
+		});
+
+		//resubmit
+		{
+			let mut resubmit_transactions = Vec::new();
+
+			for retracted in tree_route.retracted() {
+				let hash = retracted.hash;
+
+				let block_transactions = api
+					.block_body(hash)
+					.await
+					.unwrap_or_else(|e| {
+						log::warn!(target: LOG_TARGET, "Failed to fetch block body: {}", e);
+						None
+					})
+					.unwrap_or_default()
+					.into_iter()
+					.filter(|tx| tx.is_signed().unwrap_or(true));
+
+				let mut resubmitted_to_report = 0;
+
+				resubmit_transactions.extend(
```

Comment:
I think here we could encounter a situation where we previously already submitted blocks from the mempool in `update_view`, which we now find again in the retracted blocks. We should filter whether they are already in the view. Not critical, but there would be some useless revalidation.

Reply:
Please cross-check, but from what I see it will be filtered before validation, here:
https://github.com/paritytech/polkadot-sdk/blob/13cd7419f9d8d8dbc0aa429e50a995dbf1de8779/substrate/client/transaction-pool/src/graph/pool.rs#L416